presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java

```diff
@@ -1691,10 +1691,6 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         HiveTableHandle handle = (HiveTableHandle) tableHandle;
         checkArgument(!handle.getAnalyzePartitionValues().isPresent() || constraint.getSummary().isAll(), "Analyze should not have a constraint");
 
-        if (handle.getPartitions().isPresent()) {
-            return Optional.empty(); // TODO: optimize multiple calls to applyFilter
```

Comment:
Alternative (or complementary) approach would be to "squash" all the filters before invoking `applyFilter` for the first time.

Reply:
Yes, although that's tricky. It'd require running PredicatePushdown before all other rules run, but PPD benefits from other things such as unaliasSymbols, etc, and all the other rules might benefit from the result of applyFilter...

I was thinking we may need a different approach, but haven't settled on what, exactly. Some options:
1. Skip trying to enumerate partitions in the connector if there are no predicates over partition keys. This could limit the ability of predicate inference to use that info for propagating to the other side of a join, though, especially when there are only a few partitions.
2. Try enumerating partitions up to a point. If there are more than a certain threshold, stop and say "ALL" (and record that fact to avoid doing it all over again unless a different filter is provided). This still can require calling the metastore multiple times if there multiple filters stack on each other, but at least can avoid enumerating all partitions.

Reply:
>  especially when there are only a few partitions. 

is this often the case?

> Try enumerating partitions up to a point. If there are more than a certain threshold, stop and say "ALL"

We've seen that in some cases, the call for partitions (even with _some_ filter) can take very long time (>20s).
The reason is that the call translates to metastore db query containing `partitonName like `partkey1=%/partkey2=x` (when only `partkey2 is provided).
Metastore db can decide that filter is not filtering enough (the constant prefix is too short to use an index) and go full-table-scan over partitions.  🕤

Of course, there are other APIs we could call and https://github.com/prestosql/presto/issues/611 is all about that. But it won't be solved tomorrow.

Now, in some cases we may need to repeat a call. Or, do we? Maybe if we got part list once, we can only fine-tune the selection in memory, without a trip to HMS.
Regardless, since HMS can in some cases have a big problem answering the call, it's important to derive as much information as possible from the query, before making first call.

> It'd require running PredicatePushdown before all other rules run, but PPD benefits from other things such as unaliasSymbols, etc, and all the other rules might benefit from the result of applyFilter...

That's true. Part list for one joined table can result in some additional conditions for the other table.
Anyway, we should derived as much information from the query itself before making first trip to HMS.
Think of it as a preparatory PPD (a bit like "dry-run"), then HMS, then PPD to distribute the information. (Then, maybe, refine part selection, this time without calling HMS.)

Reply:
> Think of it as a preparatory PPD (a bit like "dry-run"), then HMS, then PPD to distribute the information. (Then, maybe, refine part selection, this time without calling HMS.)

Yes, that would be *somewhat* like option 1 above. Don't call the HMS until there's at least one filter over a partition key. But it's not exactly the same as what you describe because there are no "stages" for the optimization process. There's no way for the connector (or engine) to know that nothing else will be pushed down close to the TableScan.

As a workaround, for now, I'm looking into adding a call to PredicatePushdown above all the other optimizers. This should at least cause all the filters that can be squashed down in a single pass to be moved close to the table scan. We need to think about how this works in the long term, but #611 looks promising.

Reply:
> As a workaround, for now, I'm looking into adding a call to PredicatePushdown above all the other optimizers

Ugh, and that doesn't work because PPD changes the plan enough to make decorrelation rules fail.

Reply:
> Now, in some cases we may need to repeat a call. Or, do we? Maybe if we got part list once, we can only fine-tune the selection in memory, without a trip to HMS.

Subsequent `applyFilter` calls only restrict the filter, never widen it, so the initial list of partitions can only become smaller (or unchanged).

Reply:
> Subsequent `applyFilter` calls only restrict the filter, never widen it, so the initial list of partitions can only become smaller (or unchanged).

Right. But we still need to ensure we gather as many filters as possible to avoid overloading coordinator (and HMS) when doing this initial listing (a.k.a. hitting `hive.max-partitions-per-scan` limit).

Reply:
> Ugh, and that doesn't work because PPD changes the plan enough to make decorrelation rules fail.

@martin Could you give an example of that? It seems that PPD could provide better filters so that decorrelation could be even more efficient. For example, pushing equality conditions to lateral join.